File: codes/Question5_2/Java/DbDLinkedList/DbDLinkedList.java
package DbDLinkedList;

public class DbDLinkedList<T> implements DbDLinkedListInterface<T>{
    private Node<T> head;
    private Node<T> tail;
    private Node<T> cur;
    private int numOfData;

    public DbDLinkedList() {
        this.cur = null;
        this.numOfData = 0;

        Node<T> hDummyNode = new Node<T>();
        Node<T> tDummyNode = new Node<T>();

        this.head = hDummyNode;
        this.tail = hDummyNode;

        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    @Override
    public void LInsert(T data) {
        Node<T> newNode = new Node<T>();
        newNode.data = data;

        newNode.prev = this.tail.prev;
        newNode.next = this.tail;

        this.tail.prev.next = newNode;
        this.tail.prev = newNode;

        this.numOfData++;
    }

    @Override
    public boolean LFirst(T[] data) {
        if (this.tail.prev == this.head)
            return false;

        this.cur = this.head.next;
        data[0] = this.cur.data;

        return true;
    }

    @Override
    public boolean LNext(T[] data) {
        if (this.cur.next == this.tail)
            return false;

        this.cur = this.cur.next;
        data[0] = this.cur.data;

        return true;
    }

    @Override
    public T LRemove() {
        Node<T> rpos = new Node<T>();
        T rdata = this.cur.data;

        this.cur.prev.next = this.cur.next;
        this.cur.next.prev = this.cur.prev;

        this.cur = this.cur.prev;
        this.numOfData--;
        return rdata;
    }

    @Override
    public int LCount() {
        return this.numOfData;
    }
}

File: codes/Question5_2/C/DbDLinkedList.c
#include <stdio.h>
#include <stdlib.h>
#include "DbDLinkedList.h"

void ListInit(List* plist)
{
	Node* hDummyNode = (Node*)malloc(sizeof(Node));
	Node* tDummyNode = (Node*)malloc(sizeof(Node));

	plist->head = hDummyNode;
	plist->tail = tDummyNode;

	plist->head->next = plist->tail;
	plist->tail->prev = plist->head;
}

void LInsert(List* plist, Data data)
{
	Node* newNode = (Node*)malloc(sizeof(Node));
	newNode->data = data;

	newNode->prev = plist->tail->prev;
	newNode->next = plist->tail;

	plist->tail->prev->next = newNode;
	plist->tail->prev = newNode;

	(plist->numOfData)++;
}

int LFirst(List* plist, Data* pdata)
{
	if (plist->tail->prev == plist->head)
		return FALSE;

	plist->cur = plist->head->next;
	*pdata = plist->cur->data;

	return TRUE;
}

int LNext(List* plist, Data* pdata)
{
	if (plist->cur->next == plist->tail)
		return FALSE;

	plist->cur = plist->cur->next;
	*pdata = plist->cur->data;

	return TRUE;
}

Data LRemove(List* plist)
{
	Node* rpos = plist->cur;
	Data rdata = plist->cur->data;

	plist->cur->prev->next = plist->cur->next;
	plist->cur->next->prev = plist->cur->prev;

	plist->cur = plist->cur->prev;
	(plist->numOfData)--;
	return rdata;
}

int LCount(List* plist)
{
	return plist->numOfData;
}